### PR TITLE
Proposal for SIG OpenTelemetry on Mainframe

### LINKED ISSUE #1686: Proposal for establishing the SIG OpenTelemetry on Mainframe
# Description

As we see a growing interest from customers to use OpenTelemetry on the mainframe, we'd like to propose a working group to define and stabilize semantic conventions supporting the observability of the mainframe and to introduce instrumentation for programming languages present on the mainframe.

This proposal has been created as result of a [series of meetings](https://docs.google.com/document/d/14p-bpofozTL4n3jy6HZH_TKjoOXvog18G1HBRqq6liE/edit?pli=1#heading=h.pss2tdsd549w) for the topic “OpenTelemetry on Mainframe” with contributions from participants of multiple vendors and is based on [this document](https://docs.google.com/document/d/1SM5BmM4st8n4giWuqe1shqN98d37VG9C_-di1hG6sbg/edit#heading=h.q3h9nwmyzjva).

# Goals

-   Mainframe users can capture infrastructure metrics and logs with OpenTelemetry and can export these to the tools of their choice.
-   Mainframe users can capture application spans (traces), metrics, logs, and (eventually) profiles from their applications with OpenTelemetry and can export these to the tools of their choice.
    -   Context propagation works across different parts (subsystems) of each mainframe application, akin to how requests are traced across distributed services running on hosts and containers.
-   All captured telemetry has appropriate resource and interaction metadata attached to it, and this metadata conforms to the OpenTelemetry semantic conventions.
-   Capturing infrastructure and application telemetry from a mainframe incurs a minimal performance impact and allows for different deployment scenarios of the OpenTelemetry Collector to minimize both mainframe CPU costs and mainframe network egress costs.

# Staffing / Help wanted

This group will welcome contributors from various areas:
-   We look for participation from mainframe experts of all kinds: observability and monitoring experts, performance experts, virtualization experts, architects, and users of observability solutions for the mainframe and those who want to extend, simplify, and improve their current observability solution to the mainframe.
-   We welcome developers working with Assembler, COBOL, PL/1, and REXX, and engineers willing to contribute with the instrumentation for mainframe-specific programming languages.
-   We invite everybody to join who is passionate about the mainframe.
-   Two project leads who are willing to drive the project and address any issues which are not handled by other project members.
-   At least two sponsoring TC (Technical Committee) members. TC sponsors are dedicated to attending meetings, reviewing proposals, and in general being aware of the state of the project and its technical details. TC sponsors guide the project through the spec process, keep the tracking issue up to date, and help to ensure that relevant community members provide input at the appropriate times.
-   Contributors from Semantic Conventions WG: \<to be added\>
-   Maintainers or approvers: \<to be added\>

# Deliverables

-   Definition of semantic conventions that capture the concepts of the mainframe including hardware, virtualization, z/OS and subsystems, Linux on IBM Z by aligning to and extending the existing semantic conventions across resources, metrics, logs, and traces. The enhancements of the semantic conventions are assumed to be generically applicable to the mainframe and apply to any OS on the mainframe incl. TPF and z/VSE.
-   Alignment of semantic conventions for virtualization between mainframe and other virtualizations solutions, as far as feasible and reasonable
-   Description of mapping between names and mainframe terms (for conventions not specifically defined as mainframe extensions)
-   Definition of a curated, most meaningful set of metrics (out of the thousands of available metrics) to be fully supported by the semantic conventions.
-   Description and definition of semantic conventions for metrics determined as result from aggregation across system topology to simplify handling in observability backends
-   Definition of approach for supporting instrumentation of mainframe-specific programming languages
-   Definition of approach for supporting instrumentation of mainframe-specific logs
-   Definition of approach for supporting trace context propagation
-   Code instrumentation (SDKs) for mainframe-specific programming languages and runtimes

# Logistics

Start of workgroup at the beginning of October 2023.

## Track: Semantic Conventions
-   Stage 1: Semantic conventions for basic mainframe concepts - 4Q2023/1Q2024
-   Stage 2: Semantic conventions for curated list of metrics - 2Q2024

## Track: Code Instrumentation
-   Stage 1: Definition of approach for supporting instrumentation of mainframe-specific programming languages - 4Q2023.
-   Stage 2: SDK for COBOL - 1H2024
-   Stage 3: others

## Track: Collector Enhancements
-   Stage 1: Make the Collector compatible (runnable) within mainframe environments. Should be able to run from Linux on IBM Z, the Container Extension of z/OS (zCX) and the z/OS Unix System Services (USS). (Note: s390x support for Linux on IBM Z and zCX is already addressed by this [PR](https://github.com/open-telemetry/opentelemetry-collector-releases/pull/384)).
-   Stage 2: Make the Collector capture resource metrics from mainframes for z/OS and Linux on IBM Z, considering both agent and gateway deployments of the Collector.
-   Stage 3: Make the Collector capture logs from mainframes for z/OS and Linux on IBM Z, considering both agent and gateway deployments of the Collector.

## Project Leads
[\@rrschulze](https://github.com/rrschulze) agrees to volunteer, at least one more volunteer is required.
[Aaron Young](mailto:aaron.young@broadcom.com) from Broadcom is willing based on specific tasks and time requirements.

## Proposed Time 
-   For now, proposing: Tuesdays or Wednesdays at 10 a.m. PDT.
-   Please vote or please propose alternative times!
    -   Aaron Young - Either will work for me. Tuesday is slightly preferable.

## Slack Channel
[https://cloud-native.slack.com/archives/C05PXDFTCPJ](https://cloud-native.slack.com/archives/C05PXDFTCPJ)

# Examples: Enhancements of Semantic Conventions

## Metrics examples

[Metrics Semantic Conventions](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/)

| **Metrics**   | **Enhancements**  | **OS**           |
|------------- |-----------------|--------------|
| [system.cpu. - Processor metrics](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/system-metrics/)    | Multiple types of processor (e.g., CP, IFL,..) <br /> Support of partitioning and virtualization (LPAR, VM)                                                                                                                               | z/OS, Linux |
| [system.memory. - Memory metrics](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/system-metrics/)  | Several types of storage = memory, support of detailed metrics                                                                                                             | z/OS  |
| [process. - Process metrics](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/process-metrics/)                                                              | Similar to z/OS address space but different <br /> Support of Controll Block metrics  | z/OS and z/OS USS |
| [hw. - Common hardware metrics](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/hardware-metrics/) | Detailed level mapping required (e.g., add type=crypto) | z/OS, Linux       |
| [hw.cpu. - Physical processor metrics](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/hardware-metrics/)  | Multiple types of processors (e.g., CP, IFL, ...) <br /> Support of physical processor utilization| z/OS, Linux  |
| [hw.host. - Physical host metrics](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/hardware-metrics/) | For instance, LPAR power metrics  | z/OS and Linux    |
|      | Support of I/O Channels, FC, DASD, Crypto | z/OS, Linux       |
|      | Sysplex & CF, Workload Management    | z/OS              |
|      | Subsystem support | z/OS              |
| [Hostmetrics receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/hostmetricsreceiver/internal/scraper/processscraper/documentation.md) <br /> process.command <br /> process.command_line  <br /> process.executable.name  <br /> process.executable.path | Because of concept differences, sometimes cannot be fully conformed. If we match process as z/OS address space, it's hard to follow exactly since the terms are different. | z/OS  |

## Resource examples

[Resource Semantic Conventions](https://opentelemetry.io/docs/specs/otel/resource/semantic_conventions/)

| **Metrics**   | **Enhancements**  | **OS**           |
|------------- |-----------------|--------------|
|  [service. - Service metrics](https://opentelemetry.io/docs/specs/otel/resource/semantic_conventions/)  | How to describe where the data is originally created?         |      z/OS, Linux |
| | How to describe resource relationships and relationships between infrastructure components (host to LPARs to VMs)?   | z/OS, Linux |

## Log Attribute examples

[Log Attribute Semantic Conventions](https://opentelemetry.io/docs/specs/otel/logs/semantic_conventions/)

| **Metrics**   | **Enhancements**  | **OS**           |
|------------- |-----------------|--------------|
| [event. - Event attributes](https://opentelemetry.io/docs/specs/otel/logs/semantic_conventions/events/)  |  Many other attributes other than the suggested one, e.g., for correlation to other signal types  | z/OS |

### z/OS syslog Write-to-Operator (WTO) example
zos_logs3

### PR DESCRIPTION
**Pull request by rrschulze**
This is the proposal for a work group and project focusing on enhancing OpenTelemetry to support the Mainframe. For details, please see the proposal in projects/mainframe.md.

Resolves #1686.

### CODE REVIEW

**Comment by tedsuo** [comment]
Hi @rrschulze. Can you include the standard staffing section from the template? https://github.com/open-telemetry/community/blob/main/project-template.md#staffing--help-wanted

File: projects/mainframe.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+-   We welcome developers working with Assembler, COBOL, PL/1, and REXX, and engineers willing to contribute with the instrumentation for mainframe-specific programming languages.
+-   We invite everybody to join who is passionate about the mainframe.
+-   Two project leads who are willing to drive the project and address any issues which are not handled by other project members.
+-   At least two sponsoring TC (Technical Committee) members. TC sponsors are dedicated to attending meetings, reviewing proposals, and in general being aware of the state of the project and its technical details. TC sponsors guide the project through the spec process, keep the tracking issue up to date, and help to ensure that relevant community members provide input at the appropriate times.
+-   Contributors from Semantic Conventions WG: \<to be added\>
+-   Maintainers or approvers: \<to be added\>
```
Comment (tigrannajaryan) [inline projects/mainframe.md:38]: I think the new process requires that the staffing is figured out before the [project is started](https://github.com/open-telemetry/community/blob/main/project-template.md#required-staffing). From this doc:

>Projects cannot be started until the following participants have been identified

It would be useful to start with finding TC sponsors since I know most TC members are already sponsoring other projects and likely won't be available.

Comment (mtwo) [reply projects/mainframe.md:38]: That's correct @tigrannajaryan. We have plenty of engineers from IBM and other firms who are willing to work on this, but we'll need some level of TC engagement. @dyladan, @alolita, and myself are all from the GC and we're part of this group. I don't expect this project to need as much TC attention as others that require spec changes like the semantic SIGs or RUM (which required new semantics for new scenarios).

I'm reaching out to the TC now. We weren't certain if the issue was meant to be filed at the same time as contacting the TC, or after.

**Comment by rrschulze** [comment]
I confirm to contribute to this SIG.

**Comment by jackjia-ibm** [comment]
I confirm to contribute to this SIG. I’m Jack T. Jia and currently working for IBM.

**Comment by v1gnesh** [comment]
Hello @rrschulze, could you please invite me to the mentioned slack channel.
I may not be able to attend the meetings due to time zone differences, however, can follow along from notes & recordings.

PS: I'm a mainframer representing only myself.

**Comment by rrschulze** [comment]
@v1gnesh it is the Slack channel [#otel-mainframes](https://cloud-native.slack.com/archives/C05PXDFTCPJ) in the [CNCF Slack workspace](https://slack.cncf.org/). If you're not yet a member of the Slack workspace, you'll need to join it first and then you can become a member of the channel #otel-mainframes yourself.

**Comment by v1gnesh** [comment]
It seems to be... broken? https://github.com/openmainframeproject/tac/issues/535#issuecomment-1777177352

**Comment by jimporell** [comment]
working through CLA discussion within Rocket, but plan to participate

**Comment by msomasu** [comment]
I confirm to contribute to this SIG. I’m Muruganandam (Anand) Somasundaram and currently working for IBM.

**Comment by jimporell** [comment]
committed. Working on getting CLA signed

**Comment by trask** [comment]
> @rrschulze Am I still supposed to see the cert warning in the [slack workspace](https://slack.cncf.org/)? Also, can I sign up with email rather than GH/Google, etc. ?

what cert warning are you seeing?

you should be able to sign up with any email you wish

**Comment by davdai01** [comment]
I confirm to contribute to this SIG. - David Dai, Rocket Software

**Comment by v1gnesh** [comment]
@trask 
image

**Comment by jmertic** [comment]
> @trask image

@dyladan - what is the right escalation path for issues accessing CNCF Slack? https://servicedesk.cncf.io/?

**Comment by dyladan** [comment]
> > @trask image
> 
> @dyladan - what is the right escalation path for issues accessing CNCF Slack? https://servicedesk.cncf.io/?

That would be my best guess

**Comment by MacNale** [comment]
I confirm to contribute to this SIG - Machhindra Nale @ Broadcom

**Comment by rrschulze** [comment]
Hi @tedsuo, @mtwo can you please advise what needs to happen next for this proposal?

**Comment by mtwo** [comment]
Reconciling meeting updates and discussions with this issue. [We discussed this in the October GC / TC meeting](https://docs.google.com/document/d/1jylE5uZCKV9mrPw8Qrc5ExGyRVbBdqcbWPwni-hB5dE/edit#bookmark=id.qzfx4mtz8eus).

The TC is supportive of the mainframe effort given that it doesn’t stretch existing OTel technical leadership, who are already slammed with semantic convention stabilization. This means:
- No major semantic changes are required in the short-term, the focus should be on fitting mainframe telemetry metadata into the existing OTel semantics
- There are net-new people to OTel leading this (confirmed via the comments above)

I will be the GC sponsor for this project, and I think that @alolita and @dyladan plan to participate to some extent.

**Review by mtwo — APPROVED** [review]

**Comment by rrschulze** [comment]
Thanks @mtwo. @alolita or @dyladan, may I ask you to do a review of the PR as well?

**Comment by rrschulze** [comment]
We need at least one more review and approval. Who can help with it?

**Comment by jmertic** [comment]
Let me know when the meeting schedule is set; I want to cross-post to the Open Mainframe Project Community calendar

File: projects/mainframe.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+-   We look for participation from mainframe experts of all kinds: observability and monitoring experts, performance experts, virtualization experts, architects, and users of observability solutions for the mainframe and those who want to extend, simplify, and improve their current observability solution to the mainframe.
+-   We welcome developers working with Assembler, COBOL, PL/1, and REXX, and engineers willing to contribute with the instrumentation for mainframe-specific programming languages.
+-   We invite everybody to join who is passionate about the mainframe.
+-   Two project leads who are willing to drive the project and address any issues which are not handled by other project members.
+-   At least two sponsoring TC (Technical Committee) members. TC sponsors are dedicated to attending meetings, reviewing proposals, and in general being aware of the state of the project and its technical details. TC sponsors guide the project through the spec process, keep the tracking issue up to date, and help to ensure that relevant community members provide input at the appropriate times.
```
Comment (trask) [inline projects/mainframe.md:36]: can you add the TC/GC sponsors to the staffing section below?

**Comment by trask** [comment]
can you fix the spell checking CI by adding the unknown words to https://github.com/open-telemetry/community/blob/main/.cspell.yaml?

**Comment by rrschulze** [comment]
@jmertic we will start the weekly meetings for the SIG OpenTelemetry on Mainframe next Tuesday, January 16, at 10:00 am PT. Please see the [OpenTelemetry calendar](https://calendar.google.com/calendar/u/0/embed?src=google.com_b79e3e90j7bbsa2n2p5an5lf60@group.calendar.google.com).

File: projects/mainframe.md (hunk 2 above)
Comment (rrschulze) [reply projects/mainframe.md:36]: Added GC sponsors, and statement with respect to TC support.

File: projects/mainframe.md (hunk 1 above)
Comment (rrschulze) [reply projects/mainframe.md:38]: Added GC sponsors, and statement with respect to TC support.

**Comment by trask** [comment]
@dyladan @alolita can you review / approve since you are listed as sponsors? thx

**Comment by alolita** [comment]
Hi Folks! Sorry about the delayed response. Thrilled to see this SIG for OpenTelemetry on mainframe platforms take off. I plan to participate in the discussions and reviews. Also I support @mtwo in sponsoring this SIG.

**Review by alolita — APPROVED** [review]
Lgtm. Thanks for the suggestions @trask @tigrannajaryan
